refactor!: remove generic parameter from proof plans and exprs

JayWhite2357 opened the pull request:
# Rationale for this change

The `DynProofPlan` type does not need to have a generic parameter. It only has one as a relic of past requirements.

# What changes are included in this PR?

The first four commits remove the generic `Scalar` parameter from `LiteralValue`. In order to not use `Scalar` as the underlying store for the `Decimal75` variant, an `I256` type is introduced. (This type could be extended to become the main 256-bit integer used in the crate, but this is left for later work.)

The next 3 commits are the bulk of the PR and remove the generic parameter from the `ProofPlan` and `ProofExpr` traits as well as all implementers of them, such as `DynProofPlan` and `DynProofExpr`.

# Are these changes tested?
Yes. This is a refactoring change, and existing test pass.

iajoiner commented on crates/proof-of-sql/src/base/database/literal_value.rs:
```diff
@@ -63,7 +63,8 @@ impl<S: Scalar> LiteralValue<S> {
             Self::Int(i) => i.into(),
             Self::BigInt(i) => i.into(),
-            Self::VarChar((_, s)) | Self::Decimal75(_, _, s) => *s,
+            Self::VarChar(s) => s.into(),
```
NIT: I'd probably mark the `s` as `str` just to separate it from scalars.

iajoiner commented on crates/proof-of-sql/src/sql/proof_exprs/test_utility.rs:
```diff
@@ -117,9 +117,9 @@ pub fn const_varchar<C: Commitment>(val: &str) -> DynProofExpr<C> {
 
 /// Create a constant scalar value. Used if we don't want to specify column types.
-#[allow(dead_code)]
+#[cfg(test)]
```
Hmm that's not a test itself

iajoiner commented on crates/proof-of-sql/src/base/math/i256.rs:
```diff
@@ -0,0 +1,50 @@
+impl From<num_bigint::BigInt> for I256 {
+    fn from(value: num_bigint::BigInt) -> Self {
+        let (sign, limbs_vec) = value.to_u64_digits();
+        let num_limbs = limbs_vec.len().min(4);
```
We probably need to note explicitly that we are truncating

JayWhite2357 commented:
I addressed the review, fixed a bug, added tests to i256, and ran cargo fmt.

I put all of these changes in one commit: https://github.com/spaceandtimelabs/sxt-proof-of-sql/pull/339/commits/e1372060ec479fa5c7f7f3ed151939f5be7b53cb. I figured this will be the easiest way to review.

I will rebase and clean up the commits before merging. If you want me to rebase now before you review, let me know.

iajoiner commented on crates/proof-of-sql/src/base/math/i256.rs:
```diff
@@ -17,34 +17,252 @@ impl I256 {
+
+    #[must_use]
+    /// Conversion from a [`num_bigint::BigInt`].
+    /// The conversion handles negative values and also wraps when the value is to large for an `I256`.
```
too

iajoiner reviewed (approved):
Really thanks! Merge on green.
